### PR TITLE
HDF5 reader and writer of CrystalMap objects

### LINKED ISSUE #59: Reader/writer for CrystalMap
I want a simple HDF5 reader/writer for the `CrystalMap` class into v0.3 (#58), so we can start to put stuff from pyxem in it, and read the data back into pyxem and kikuchipy.

I like the h5ebsd format specified here: Jackson, M.A., Groeber, M.A., Uchic, M.D. et al. h5ebsd: an archival data format for electron back-scatter diffraction data sets. Integrating Materials 3, 44–55 (2014). https://doi.org/10.1186/2193-9772-3-4. I plan to follow this as closely as it suits our `CrystalMap` class. Will add `manufacturer` and `version` datasets in the top group. This way we can easily create readers into e.g. MTEX's EBSD.load of this format.

Are there any interest in having readers returning anything else than a `CrystalMap`? If not, we should consider having only `orix.load()` figuring out which format the file is in, and returning a `CrystalMap` using the proper reader (like HyperSpy's clever `hs.load()`).

I would like to have the possibility to use the `Phase` and `PhaseList` classes in kikuchipy, not always via a `CrystalMap`. I therefore plan to make the reader/writer modular, so that these two classes are written to more or less standard data groups and sets.

Does anyone have any other requirements for the reader/writer? I'm not familiar with everything pyxem would want to write, so if there are anything that does not fit into a `CrystalMap`, please let me know.

### PR DESCRIPTION
**Pull request by hakonanes**
Signed-off-by: Håkon Wiik Ånes <hwaanes@gmail.com>

Closes #59.

Adds a master load function, `orix.io.load()`, which takes a filename and keyword arguments, determines the file format from the file extension and calls the correct reader. Readers have been moved to `orix.io.plugins` and their `load_x` function have been standardized to `file_reader`. This IO format is adopted from HyperSpy.

Todo:
- [x] Merge #66 into this branch *after* it is merged into master.
- [x] Reader/writer of `CrystalMap` objects to HDF5 file
̃- [ ] ~Reader/writer of `Phase` objects to HDF5 group (to be included within an HDF5 file)~ To/from dictionary instead of to/from HDF5 group
- [x] ~Reader/writer of `PhaseList` objects to HDF5 group (to be included within an HDF5 file)~ To/from dictionary instead of to/from HDF5 group
- [x] Master save function, to be used by the different classes via their own `.save()`.
- [x] Tests.

Want modular reader/writer so that classes' readers/writers can be used in other packages.

### CODE REVIEW

**Comment by dnjohnstone** [comment]
Thanks for this @hakonanes - will be an improvement.

Do we want it to be orix.load rather than orix.io.load though?

**Comment by hakonanes** [comment]
I placed it in orix.io.load instead of orix.load because so far, no modules are importable from `orix/__init__.py` (the file only contains release info). I agree that orix.load would be better.

I think the import structure in general should be changed. E.g., if we only want the class `Rotation` to be importable from the `rotation.py` file, we should make it available via `orix.quaternion.Rotation` instead of `orix.quaternion.rotation.Rotation`, which is how it is now. If we want more control over the API, we could use the `__all__` list in `__init__.py`, like e.g. scikit-image does (https://github.com/scikit-image/scikit-image/blob/master/skimage/exposure/__init__.py)

**Comment by pc494** [comment]
I would be happy to see the import style change. Would it be a huge hassle to do it in a seperate PR after all the `0.3.0` functionality is in?

**Comment by hakonanes** [comment]
Glad we agree! No, we can do this then.

**Comment by hakonanes** [comment]
Have a minor issue with initializing a crystal map with a `phase_id` array containing e.g. [0, 1] unique phase IDs, while passing a `phase_list` with IDs [1, 2]. Ready for review after this is fixed.

**Comment by pc494** [comment]
@hakonanes this is good to go for review?

**Comment by hakonanes** [comment]
Sorry, I had to correct some explanations in plugins' `file_reader` docstring, since they return `CrystalMap` objects directly and not dictionaries to create the objects.

**Comment by pc494** [comment]
Making progress on my review;  am I right in thinking that all three readers are called `file_reader` under the assumption a user will only ever be working with one form of data(?), and by extension that we don't mind the naming conflicts if more than 1 are imported?

**Comment by hakonanes** [comment]
Great, thanks for reviewing so quickly!

No... The setup is identical to HyperSpy's load functionality. The user should only call orix.io.load(), which looks through all the plugins, finds the correct one based on the file extension (and possible HDF5 footprints, like "crystal_map" among the top HDF5 groups), and then calls the plugin's file_reader(). The orix.io.save() function operates the same way.

A drawback to this is that all possible arguments in the file_reader/file_writer functions must either be documented in the save/load functions, or we can ask the user to look up the docstrings of each plugin's file_reader(). In the latter case this information should be easily accessible (like it is for HyperSpy: http://hyperspy.org/hyperspy-doc/current/api/hyperspy.io.html#hyperspy.io.load), e.g. in the form of docs via readthedocs. I think we should provide this, at least the API which is autogenerated from the docstrings by sphinx.

**Comment by hakonanes** [comment]
See HyperSpy's IO plugin readme for further details: https://github.com/hyperspy/hyperspy/tree/RELEASE_next_minor/hyperspy/io_plugins.

**Comment by pc494** [comment]
> Great, thanks for reviewing so quickly!
> 
> ~No...~ Sort of. The setup is identical to HyperSpy's load functionality. The user should only call orix.io.load(), which looks through all the plugins, finds the correct one based on the file extension (and possible HDF5 footprints, like "crystal_map" among the top HDF5 groups), and then calls the plugin's file_reader(). The orix.io.save() function operates the same way.
> 
> A drawback to this is that all possible arguments in the file_reader/file_writer functions must either be documented in the save/load functions, or we can ask the user to look up the docstrings of each plugin's file_reader(). In the latter case this information should be easily accessible (like it is for HyperSpy: http://hyperspy.org/hyperspy-doc/current/api/hyperspy.io.html#hyperspy.io.load), e.g. in the form of docs via readthedocs. I think we should provide this, at least the API which is autogenerated from the docstrings by sphinx.

The curse of reviewing code out of order without having seen all the files first. This method is my preference.

**Comment by hakonanes** [comment]
It's my fault, I should have provided this information first.

File: orix/crystal_map/phase_list.py [hunk 1]
```diff
@@ -298,20 +298,24 @@ def __init__(
...
+            if ids is None:
+                ids = 0
+            d = {ids: phases}
         else:
             # Ensure possible single strings or single objects have
             # iterables of length 1
             if isinstance(names, str):
                 names = list((names,))
-            if isinstance(symmetries, str) or isinstance(symmetries, Symmetry):
+            if (
```
[the remark this answers is not in the document]
Comment (hakonanes) [reply orix/crystal_map/phase_list.py]: You're right. Considering how fond I am of that function, I should use it better... I looked at all other uses and updated the logic where possible. Thanks!

**Pull request merged**